File: packages/cli/src/commands/skills.ts
import { execSync } from "node:child_process";
import { Command } from "commander";
import { ApiClient } from "../lib/api-client.js";
import {
  printJson,
  printTable,
  success,
  error as printError,
  info,
  fmt,
} from "../lib/output.js";

export function skillsCommand(): Command {
  const cmd = new Command("skills").description("Browse and install skills");

  cmd
    .command("list")
    .description("List available skills")
    .option("--layer <layer>", "Filter by layer ID")
    .option("--json", "Output as JSON")
    .action(async (opts) => {
      const client = new ApiClient(cmd.optsWithGlobals());
      const layers = await client.listSkills(opts.layer);

      if (opts.json) {
        printJson(layers);
        return;
      }

      for (const layer of layers) {
        console.log(`\n${fmt.bold(fmt.cyan(layer.name))} ${fmt.dim(`(${layer.id})`)}`);
        console.log(fmt.dim(layer.description));

        printTable(
          ["Slug", "Name", "Description"],
          layer.skills.map((s) => [s.slug, s.name, s.description]),
        );
      }
    });

  cmd
    .command("get")
    .description("Get details for a skill")
    .argument("<slug>", "Skill slug")
    .option("--json", "Output as JSON")
    .action(async (slug: string, opts) => {
      const client = new ApiClient(cmd.optsWithGlobals());
      const skill = await client.getSkill(slug);

      if (opts.json) {
        printJson(skill);
        return;
      }

      console.log(fmt.bold(fmt.cyan(skill.name)));
      console.log(fmt.dim(skill.description));
      console.log(`\nLayer: ${skill.layer}`);
      console.log(`Install: ${fmt.green(skill.installCommand)}`);
      console.log(`URL: ${skill.skillsUrl}`);
    });

  cmd
    .command("install")
    .description("Install a skill (delegates to npx skills add)")
    .argument("<slug>", "Skill slug")
    .action(async (slug: string) => {
      const client = new ApiClient(cmd.optsWithGlobals());
      let installCmd: string;

      try {
        const skill = await client.getSkill(slug);
        installCmd = skill.installCommand;
      } catch {
        // Fallback if API unavailable
        installCmd = `npx skills add broomva/${slug}`;
      }

      info(`Running: ${installCmd}`);
      try {
        execSync(installCmd, { stdio: "inherit" });
        success(`Installed skill: ${slug}`);
      } catch {
        printError(`Failed to install skill: ${slug}`);
        process.exit(1);
      }
    });

  return cmd;
}

File: crates/broomva-cli/src/cli/skills.rs
use crate::api::BroomvaClient;
use crate::cli::output::{OutputFormat, print_json, print_kv, print_table};
use crate::error::BroomvaResult;

pub async fn handle_list(
    client: &BroomvaClient,
    layer: Option<&str>,
    format: OutputFormat,
) -> BroomvaResult<()> {
    let skills = client.list_skills(layer).await?;

    if format == OutputFormat::Json {
        print_json(&skills);
        return Ok(());
    }

    let rows: Vec<Vec<String>> = skills
        .iter()
        .map(|s| {
            vec![
                s.slug.clone(),
                s.name.clone(),
                s.layer.clone().unwrap_or_default(),
                s.description.clone().unwrap_or_default(),
            ]
        })
        .collect();

    print_table(&["slug", "name", "layer", "description"], &rows, format);
    Ok(())
}

pub async fn handle_get(
    client: &BroomvaClient,
    slug: &str,
    format: OutputFormat,
) -> BroomvaResult<()> {
    let skill = client.get_skill(slug).await?;

    if format == OutputFormat::Json {
        print_json(&skill);
        return Ok(());
    }

    print_kv("Name", &skill.name);
    print_kv("Slug", &skill.slug);
    if let Some(ref l) = skill.layer {
        print_kv("Layer", l);
    }
    if let Some(ref d) = skill.description {
        print_kv("Description", d);
    }
    if let Some(ref c) = skill.content {
        println!();
        println!("{c}");
    }
    Ok(())
}

pub async fn handle_install(client: &BroomvaClient, slug: &str) -> BroomvaResult<()> {
    let skill = client.get_skill(slug).await?;

    if let Some(ref cmd) = skill.install_command {
        println!("  Installing {slug}...");
        let status = tokio::process::Command::new("sh")
            .arg("-c")
            .arg(cmd)
            .status()
            .await?;
        if status.success() {
            println!("  Installed {slug} successfully.");
        } else {
            eprintln!("  Install command exited with status: {status}");
        }
    } else {
        println!("  No install command defined for {slug}.");
        if let Some(ref content) = skill.content {
            println!("  Skill content:");
            println!("{content}");
        }
    }
    Ok(())
}
